### ISSUE TITLE
Options for scripting against aurae

### ISSUE THREAD

**Issue by bpmooch** [aurae-runtime/aurae#35]
I'd like to propose using bash (with an aurae cli similar to `kubectl` or `buildah`) or Deno to script against aurae instead of moving forward with the Rhai implementation. I am not heavily invested in bash or Deno per se, but I am not sure that Rhai's unique properties serve the project better than other more popular approaches to scripting against API's. In short, why is Rhai better than python, javascript, bash, etc? I tried to focus my thinking on who aurae users probably will be and how that theoretical persona would want to interact with the system. In addition, I am working on a system with some of the same goals as aurae currently, and I'm sure that project also biases my opinions about aurae. 

Soooo, from the readme,

> AuraeScript follows a similar client paradigm to Kubernetes kubectl command. However, unlike Kubernetes this is not a command line tool like kubectl. AuraeScript is a fully supported programing language complete with a systems standard library. The Aurae runtime projects supports many clients, and the easiest client to get started building with is AuraeScript.

Does it make sense that the easiest client to get started building with is a scripting language users probably aren't familiar with? As a sysadmin, I would much rather reach for bash to get started hacking. I think `buildah` serves as a great model to emulate ([tool intro here](https://github.com/containers/buildah/blob/main/docs/tutorials/01-intro.md)). `buildah` replaces dockerfiles the same way auraescript is attempting to replace yaml for infra configuration. It should be simple to spin up an instance of aurae and configure it with a simple bash script or interactively from the terminal. Bash is the most intuitive answer for this. Moving past simplicity, power users can get pretty far with bash, awk, and a well maintained gnu-style cli such as `kubectl` or `buildah` before needing to reach for a beefier (and crucially, more complex to integrate and maintain) scripting language such as python, javascript, perl, or lua.

And speaking of lua, an even simpler question to ask than above is why Rhai instead of lua? I've fiddled with it a bit to mess with my neovim configuration, but I am far from an expert. From my outsider's perspective, lua seems like an active success story for small, embed-able languages.

If the scripting language is important to provide a platform for the aurae standard library, then Deno becomes a very compelling option. Deno markets itself as a v8 runtime that is secure, hackable, and embed-able. 

- The `Deno` global object (`Deno.spawn`, `Deno.test`, `Deno.exit`) can be modified to include runtime specific functionality. You can ship aurae users a full featured runtime & compiler to build their aurae scripts with.
- Javscript & Typescript is a very popular with developers these days, and might help make maintainable infrastructure code understandable to more web developers.
- Deno takes a lot of design inspiration from Go. In fact the standard library is loosely modeled after the Go stdlib. Using Deno gives us a lot of the upsides of using Go without some of the more prickly downsides (see bottom)

## A Go Problemo

> side note: when building docker we were struggling to provide flexibly replaceable extensions as Go had no support for dynamic loading of libraries. As Docker wanted to ship with "batteries included but changeable", we ended up with those CNI / CSI constructs, where core functionality was pushed to external processes with more or less nice interfaces.

**Comment by dominichamon** [comment]
_if_ the language is meant to wrap around the gRPC API, then either a custom language or a gRPC-supported language are likely options.  ie, either `auraescript` as a custom wrapper around the gRPC client to reduce verbosity/improve readability or python/go/java(!) acting as a client directly.

i'm not sure bash or lua fits the bill unless there are gRPC client implementations i'm not aware of.

but the question of "why auraescript?" is a good one to probe.

**Comment by bpmooch** [comment]
> i'm not sure bash or lua fits the bill unless there are gRPC client implementations i'm not aware of.

Sorry for not being clear - I'm proposing a cli similar to `buildah`, which has a goal of replacing dockerfiles with bash scripts. I'm basically saying that I don't see the need for auraescript because a gRPC API wrapper cli written in rust (or whatever) seems to do the job great and avoids a lot of complexity in maintaining a Rhai implementation.

If auraescript's intent was to heavily utilize the integration of the scripting language with aurae (such as [Rhai's bespoke Rust integrations](https://github.com/rhaiscript/rhai#protected-against-attacks)), then Deno seems a more compelling choice to me

**Comment by krisnova** [comment]
@bpmooch do you believe these features are mutually exclusive? I am sold on entertaining something like Deno... I also think having a CLI like buildah where we can easily use it with bash also makes sense. 

In other words... why not both?

**Comment by krisnova** [comment]
@bpmooch would you be interested in starting a deno-based auraescript? or even a buildah style CLI tool? we can see which ones end up growing if we are interested in taking the "why not both" approach?

**Comment by bpmooch** [comment]
@krisnova back at keys after vacation, yea I'm interested in working on this. I'll try to get some notes into the architecture repo soonish

**Comment by krisnova** [comment]
Hey @bpmooch we just had an interesting thought from @GANitak 

What if we added Deno "on top" of the Rhai library? 

We know we will need to build out an opinionated client in rust (something that looks for configuration files in a well known place and understands what an aurae socket is) as well as the basic client functions anyway. 

Couldn't we just map the client functions directly into Deno as well? Or at least slowly move them over in the `lib.rs` file to Deno? Unsure if you have an approach in mind? I am open to whatever you want to do and am already convinced Deno is a good move for the project.

Basically I would like to be able to move forward with testing against the daemon right now, without blocking on the Deno work.

**Comment by krisnova** [comment]
The work in https://github.com/aurae-runtime/aurae/pull/115 will close this issue as it will introduce Deno as the primary core for AuraeScript and make AuraeScript 100% compatible with TypeScript and JavaScript. 

This was a fantastic idea.

**Comment by future-highway** [comment]
Closing as of #120
